File: model-value/src/main/java/com/bnorm/barkeep/model/ComponentValue.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model;

import org.jetbrains.annotations.NotNull;

import com.google.auto.value.AutoValue;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

@AutoValue
public abstract class ComponentValue implements Component {

  public static Builder builder() {
    return new AutoValue_ComponentValue.Builder();
  }

  public static JsonAdapter<ComponentValue> jsonAdapter(Moshi moshi) {
    return new AutoValue_ComponentValue.MoshiJsonAdapter(moshi);
  }

  @Override
  public int compareTo(@NotNull Component other) {
    return Component.Companion.getCOMPARATOR().compare(this, other);
  }

  @AutoValue.Builder
  abstract static class Builder {

    abstract ComponentValue build();

    abstract Builder setIngredient(Ingredient ingredient);

    abstract Builder setMin(double min);

    abstract Builder setMax(Double max);

    abstract Builder setComponentNum(long componentNum);

    abstract Builder setOrder(long order);
  }
}

File: model-value/src/main/kotlin/com/bnorm/barkeep/model/ComponentValue.kt
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model

import com.squareup.moshi.FromJson
import com.squareup.moshi.ToJson

data class ComponentValue(override val ingredient: Ingredient? = null,
                          override val min: Double = 0.0,
                          override val max: Double? = null,
                          override val componentNum: Long = 0,
                          override val order: Long = 0) : Component

object ComponentValueAdapter {
  @ToJson fun toJson(component: Component): ComponentValue {
    return ComponentValue(component.ingredient, component.min, component.max, component.componentNum, component.order)
  }

  @FromJson fun fromJson(json: ComponentValue): Component {
    return json
  }
}
